core/math/projection.cpp

```diff
@@ -596,101 +596,236 @@ Projection Projection::inverse() const {
 }
 
 void Projection::invert() {
-	int i, j, k;
-	int pvt_i[4], pvt_j[4]; /* Locations of pivot matrix */
-	real_t pvt_val; /* Value of current pivot element */
-	real_t hold; /* Temporary storage */
-	real_t determinant = 1.0f;
-	for (k = 0; k < 4; k++) {
-		/** Locate k'th pivot element **/
-		pvt_val = columns[k][k]; /** Initialize for search **/
-		pvt_i[k] = k;
-		pvt_j[k] = k;
-		for (i = k; i < 4; i++) {
-			for (j = k; j < 4; j++) {
-				if (Math::abs(columns[i][j]) > Math::abs(pvt_val)) {
-					pvt_i[k] = i;
-					pvt_j[k] = j;
-					pvt_val = columns[i][j];
-				}
-			}
-		}
-
-		/** Product of pivots, gives determinant when finished **/
-		determinant *= pvt_val;
-		if (Math::is_zero_approx(determinant)) {
-			return; /** Matrix is singular (zero determinant). **/
-		}
-
-		/** "Interchange" rows (with sign change stuff) **/
-		i = pvt_i[k];
-		if (i != k) { /** If rows are different **/
-			for (j = 0; j < 4; j++) {
-				hold = -columns[k][j];
-				columns[k][j] = columns[i][j];
-				columns[i][j] = hold;
-			}
-		}
-
-		/** "Interchange" columns **/
-		j = pvt_j[k];
-		if (j != k) { /** If columns are different **/
-			for (i = 0; i < 4; i++) {
-				hold = -columns[i][k];
-				columns[i][k] = columns[i][j];
-				columns[i][j] = hold;
-			}
-		}
-
-		/** Divide column by minus pivot value **/
-		for (i = 0; i < 4; i++) {
-			if (i != k) {
-				columns[i][k] /= (-pvt_val);
-			}
-		}
-
-		/** Reduce the matrix **/
-		for (i = 0; i < 4; i++) {
-			hold = columns[i][k];
-			for (j = 0; j < 4; j++) {
-				if (i != k && j != k) {
-					columns[i][j] += hold * columns[k][j];
-				}
-			}
-		}
-
-		/** Divide row by pivot **/
-		for (j = 0; j < 4; j++) {
-			if (j != k) {
-				columns[k][j] /= pvt_val;
-			}
-		}
+	// Adapted from mesa's src/util/u_math.c util_invert_mat4x4
+	// on commit 0d170a81288f2427425d53965a0daa57a57b110f
+	Projection temp;
+	real_t *out = (real_t *)temp.columns;
+	real_t *m = (real_t *)columns;
+
+	real_t wtmp[4][8];
+	real_t m0, m1, m2, m3, s;
+	real_t *r0, *r1, *r2, *r3;
+
+#define MAT(m, r, c) (m)[(c) * 4 + (r)]
+#define SWAP_ROWS(a, b)   \
+	{                     \
+		real_t *_tmp = a; \
+		(a) = (b);        \
+		(b) = _tmp;       \
+	}
```

Comment:
Why not just use https://en.cppreference.com/w/cpp/algorithm/swap ?

Reply:
Godot intentionally avoids using the STL

Reply:
In that case Godot/Redot should probably provide an alternative, macro uses like that are entirely unnecessary in C++.

Reply:
> Godot intentionally avoids using the STL

I can't find a reference to the reasoning for this, do you know it?

Reply:
> > Godot intentionally avoids using the STL
> 
> I can't find a reference to the reasoning for this, do you know it?

Here. https://docs.godotengine.org/en/stable/contributing/development/cpp_usage_guidelines.html

The document states that STL is not allowed.

Reply:
It would be nice if it were possible to implement Abseil so that Godot/Redot wouldn't need to manually implement custom types. For example, instead of using Godot's custom String type versus STL's String type, they could use Abseil with Unicode support and achieve smaller binary sizes compared to using STL.